src/tools/perf/lib/uct_tests.cc

```diff
@@ -783,21 +783,48 @@ class uct_perf_test_runner {
    TEST_CASE_ALL_OSD(_perf, _case, UCT_PERF_DATA_LAYOUT_BCOPY) \
    TEST_CASE_ALL_OSD(_perf, _case, UCT_PERF_DATA_LAYOUT_ZCOPY)
 
+#define DEFINE_DISPATCH_FUNC(_case, _cmd, _type) \
+    static ucs_status_t uct_perf_dispatch_##_case(ucx_perf_context_t *perf) { \
+        TEST_CASE_ALL_DATA(perf, (_cmd, _type)); \
+        return UCS_ERR_INVALID_PARAM; \
+    }
+
+#define DISPATCH_FUNC_ENTRY(_case, _cmd, _type) uct_perf_dispatch_##_case,
+
+#define ITERATE_TEST_CASES(_macro) \
+    _macro(pingpong_am,     UCX_PERF_CMD_AM,    UCX_PERF_TEST_TYPE_PINGPONG)   \
+    _macro(pingpong_put,    UCX_PERF_CMD_PUT,   UCX_PERF_TEST_TYPE_PINGPONG)   \
+    _macro(pingpong_add,    UCX_PERF_CMD_ADD,   UCX_PERF_TEST_TYPE_PINGPONG)   \
+    _macro(stream_uni_am,   UCX_PERF_CMD_AM,    UCX_PERF_TEST_TYPE_STREAM_UNI) \
+    _macro(stream_uni_put,  UCX_PERF_CMD_PUT,   UCX_PERF_TEST_TYPE_STREAM_UNI) \
+    _macro(stream_uni_get,  UCX_PERF_CMD_GET,   UCX_PERF_TEST_TYPE_STREAM_UNI) \
+    _macro(stream_uni_add,  UCX_PERF_CMD_ADD,   UCX_PERF_TEST_TYPE_STREAM_UNI) \
+    _macro(stream_uni_fadd, UCX_PERF_CMD_FADD,  UCX_PERF_TEST_TYPE_STREAM_UNI) \
+    _macro(stream_uni_swap, UCX_PERF_CMD_SWAP,  UCX_PERF_TEST_TYPE_STREAM_UNI) \
+    _macro(stream_uni_cswap,UCX_PERF_CMD_CSWAP, UCX_PERF_TEST_TYPE_STREAM_UNI)
+
+ITERATE_TEST_CASES(DEFINE_DISPATCH_FUNC)
+
 ucs_status_t uct_perf_test_dispatch(ucx_perf_context_t *perf)
 {
-    UCS_PP_FOREACH(TEST_CASE_ALL_DATA, perf,
-        (UCX_PERF_CMD_AM,  UCX_PERF_TEST_TYPE_PINGPONG),
-        (UCX_PERF_CMD_PUT, UCX_PERF_TEST_TYPE_PINGPONG),
-        (UCX_PERF_CMD_ADD, UCX_PERF_TEST_TYPE_PINGPONG),
-        (UCX_PERF_CMD_AM,  UCX_PERF_TEST_TYPE_STREAM_UNI),
-        (UCX_PERF_CMD_PUT, UCX_PERF_TEST_TYPE_STREAM_UNI),
-        (UCX_PERF_CMD_GET, UCX_PERF_TEST_TYPE_STREAM_UNI),
-        (UCX_PERF_CMD_ADD, UCX_PERF_TEST_TYPE_STREAM_UNI),
-        (UCX_PERF_CMD_FADD, UCX_PERF_TEST_TYPE_STREAM_UNI),
-        (UCX_PERF_CMD_SWAP, UCX_PERF_TEST_TYPE_STREAM_UNI),
-        (UCX_PERF_CMD_CSWAP, UCX_PERF_TEST_TYPE_STREAM_UNI)
-        );
-
-    ucs_error("Invalid test case");
+    using uct_dispatch_func_t                      = ucs_status_t
```

Comment:
Maybe just use lambdas to keep function in a single place (no previous declaration needed):
```C++
#define DISPATCH_LAMBDA(_case, _cmd, _type) \
    [](ucx_perf_context_t *perf) { TEST_CASE_ALL_DATA(perf, (_cmd, _type)); return UCS_ERR_INVALID_PARAM; },

    uct_dispatch_func_t dispatchers[] = {
        ITERATE_TEST_CASES(DISPATCH_LAMBDA)
    };
```

Reply:
We can't use lambda here, it'll require adding lstdc++, and as discussed in https://github.com/openucx/ucx/pull/10131#discussion_r1789125666 we are trying to avoid it